### ISSUE TITLE
`isometry` and `random_isometry` — inclusion and Haar-random isometries, mostly for free

### ISSUE THREAD

**Issue by Ryo-wtnb11** [Ryo-wtnb11/TeNeT-py#89]
**Goal**

Give the library the two constructors every tensor-network algorithm opens with: the **inclusion isometry** of a smaller graded space into a larger one (initial CTM environments, gauge fixing, bond expansion with a controlled embedding), and a **Haar-random isometry** per coupled sector (random initial environments, random canonical-form MPS tensors, randomized-SVD-style probes). Today both are hand-rolled per provider or reached through a dense detour.

The headline is that half of it already exists. **`isometry` is `embed(identity(W), target legs)` and needs no new construction at all** — measured, exactly, on SU(2): single-leg and two-leg targets both give per-coupled-sector matrices whose columns are orthonormal to `0.0`, with `adjoint(iso) @ iso == identity(W)` exactly and every padded slot exactly zero. `identity` (#30) supplies the blocks, `embed` (#83) supplies the placement *and* the entire refusal surface (provider, `side`, `dual`, leg count, per-sector containment, each naming both degeneracies). What is left is a five-line composition and its documentation.

`random_isometry` is the genuinely new one: a Haar draw per coupled sector, which is a `map_layout` walk plus a QR with the Mezzadri sign fix.

**Design / API sketch**

```python
# src/tenet/ops/map.py — beside `identity`, whose signature shape they follow.

def isometry(codomain: Sequence[Leg], domain: Sequence[Leg], *,
             dtype: Any = np.float64) -> "SymmetricTensor":
    """The inclusion ``domain -> codomain``: ``W† W = id(domain)``, ``(W W†)² = W W†``.

    ``codomain[i]`` must contain ``domain[i]`` sector-wise -- same provider, same
    ``dual``, every sector of the domain present in the codomain with a degeneracy at
    least as large. ``side`` is *set*, not compared, exactly as :func:`identity` does.
    The result's legs are ``(*codomain OUT, *domain IN)``.

    The whole body, and every refusal, is ``embed`` of ``identity``::

        domain = tuple(domain)
        return embed(identity(domain, dtype=dtype),
                     (*(replace(l, side=OUT) for l in codomain),
                      *(replace(l, side=IN) for l in domain)))

    Containment is required **per leg**, which is stricter than the fused,
    sector-wise condition TensorKit's ``isometry`` imposes -- see the rationale.
    """


def random_isometry(codomain: Sequence[Leg], domain: Sequence[Leg], *,
                    seed: int | None = None, dtype: Any = np.float64) -> "SymmetricTensor":
    """A Haar-random isometry: ``W† W = id(domain)``, independent per coupled sector.

    Per coupled sector ``c``, a ``(rows_c, cols_c)`` Gaussian draw, ``qr``, and the
    sign fix ``Q * sign(diag(R))`` that makes the result Haar-distributed (Mezzadri).
    Requires ``rows_c >= cols_c`` in every coupled sector -- the *fused* containment
    condition, checked structurally from :class:`~tenet.map_view.MapLayout`, which is
    weaker than ``isometry``'s per-leg one.

    NumPy draws through ``np.random.default_rng(seed)``, exactly as
    ``SymmetricTensor.random`` (#9) and ``identity``'s ``like="numpy"`` already do:
    a constructor is not a traced operation. A complex ``dtype`` gets a genuinely
    complex (Ginibre) draw, not a real one cast -- see the rationale.
    """
```

Both exported as `tenet.isometry` / `tenet.random_isometry`, free functions beside `identity`, no `SymmetricTensor` method.

**Key decisions & rationale**

- **`isometry` reuses `embed(identity(...))` rather than building blocks, and that is the entire design.** The alternative — `from_matrices(structure, {c: ar.do("eye", rows_c, cols_c)})`, which also works and which rectangular `eye` supports on NumPy and JAX — was measured and rejected *even though it is shorter and more general*, because it names a different map. `eye(rows_c, cols_c)` sends the `j`-th column band to the `j`-th row band, which is an arbitrary correspondence whenever the two sides' band orders do not line up; `embed(identity(...))` sends each degeneracy slot of each leg to the same slot of the larger leg, which is *the* inclusion, and which agrees with the prefix convention `svd(bond=)` (#77) and `embed` (#83) already share. Reusing `embed` also inherits its refusals verbatim, which is a paragraph of error messages nobody has to write or keep consistent. Rung 2 of the ladder, and it holds.
- **Containment is required per leg, which is stricter than TensorKit — stated as a limitation with its escape named, not hidden.** TensorKit's `isometry!` checks `domain(t) ≾ codomain(t)`, i.e. containment of the *fused* spaces sector-wise, and would accept a target where no single leg contains its partner but the fusion does. Ours refuses that, because the only construction available for it is the arbitrary coupled-sector `eye` above — and TensorKit's own `isomorphism` docstring concedes the same non-canonicity ("there is no canonical choice"). Making an arbitrary basis correspondence silently, inside a function named `isometry`, is the kind of thing found six months later in a gauge bug. `ponytail:` per-leg containment; the fused version is a coupled-sector `eye` and about four lines, and it lands when a caller can say which correspondence they meant.
- **The argument order is TensorKit's `isometry(codomain, domain)` — target first — and not `identity`'s single-tuple shape, because there are two spaces.** `identity(legs)` mirrors one tuple; an isometry is a map between two, and `isometry(V, W)` reads as "into `V`, from `W`" in the same direction `Hom(W, V)` and `compose`'s `a ∘ b` do. `dtype` stays a keyword with `identity`'s default.
- **The Mezzadri sign fix is not optional and is the only subtle line in the file.** A plain `np.linalg.qr` of a Gaussian matrix is **not** Haar-distributed: LAPACK's `R` has arbitrary diagonal signs, which biases `Q`'s column phases. `Q * sign(diag(R))` fixes it, and it is the same normalization TensorKit uses (`randisometry` is `qr_compact!(randn!(rng, A); positive = true)` — the positive-`R` QR is the sign fix under another name). Measured orthogonality `4.4e-16`. This gets a comment citing Mezzadri, because it is exactly the line a future reader deletes as redundant.
- **"Haar per coupled sector" is the product of per-sector Haar measures, not Haar on the dense space, and the docstring says so.** A symmetric isometry lives in `⊕_c U(rows_c) / …`, so the block-diagonal ensemble is the *correct* one for a random symmetric tensor — but "random isometry" reads like the dense object, and someone will eventually compare a moment against a dense-Haar formula and be confused. TensorKit's `randisometry` is per-block too; naming the fact costs one sentence.
- **Complex `dtype` gets a real+imaginary Ginibre draw, deviating from `SymmetricTensor.random`'s documented shortcut on purpose.** `random`'s `ponytail:` note says "real draws cast to dtype; give complex dtypes a real+imag draw if a test ever needs genuinely complex random data". For a general random tensor that shortcut is harmless. For `random_isometry` it is a *correctness* trap: a real orthogonal matrix cast to `complex128` is an isometry, satisfies every test one would write, and is emphatically not Haar on `U(n)` — its eigenvalue statistics are those of `O(n)`. Three lines fix it, and the criteria pin it with a statistic rather than a shape check. This is the "two options, same size, take the one correct on edge cases" rule.
- **NumPy draws with `like="numpy"`, seeded through `np.random.default_rng(seed)` — `identity`'s and `SymmetricTensor.random`'s convention, unchanged.** A constructor runs at setup time, outside any trace; threading `jax.random.PRNGKey` through would put a backend branch into `ops/map.py` (which today mentions NumPy only as `identity`'s default dtype) and would make reproducibility backend-dependent. `t.to_backend("jax")` is the documented route onto a device, as it already is for `random` and `zeros`.
- **Free functions beside `identity`, not `SymmetricTensor` classmethods, and not registered in `array/dispatch.py`.** `identity` is a free function; `zeros`/`random`/`from_legs` are classmethods because they take a leg tuple and produce *that* tensor, while `identity`/`isometry` produce a *morphism* determined by a pair of spaces. #82 already settled the registration half: a constructor is not an operation on a `SymmetricTensor`, and the list stays closed (invariant 11).
- **No `unitary`, no `isomorphism`, no `random_unitary`.** TensorKit ships all three, and here they are all `isometry`/`random_isometry` with equal spaces — `unitary(V, W)` would be `isometry` plus the extra assertion that the containment is an equality, i.e. one criterion, not a second function. `isomorphism` between *different* spaces of equal sector dimensions is the arbitrary-correspondence problem again, declined for the same reason as the fused-containment case.
- **This pairs with `left_null` (filed alongside), and the pairing gets a criterion rather than a helper.** `left_null(T)` returns an isometry, so `adjoint(N) @ N == identity(bond)` is the same statement `isometry` guarantees, and a test asserting that the two agree on shape and contract is the cheapest way to keep the two definitions from drifting. No `complement_of` convenience function.

**References**

- README: "Composition remains first-class", "TensorMap views", "Fusion spaces" / "`GradedSpace`", "Recovering TensorMap semantics", "Example API"; invariants **1**, **2**, **5**, **8**, **11**.
- Code: `src/tenet/ops/map.py` (#30) — `identity`, whose signature shape, `replace(leg, side=...)` mirroring, `like="numpy"` fill and "`space`, `dual` and `name` are kept and only `side` is set" stance are all adopted here, plus `compose` and `_check_composable`; `src/tenet/ops/embed.py` (#83) — `embed` and `_check_containment`, which *are* `isometry`'s implementation and refusal surface; `src/tenet/ops/map.py` (#31) — `adjoint`, half of every criterion; `src/tenet/map_view.py` (#29) — `map_layout`, `MapLayout.shape` (the per-coupled-sector `(rows, cols)` `random_isometry` walks) and `from_matrices`; `src/tenet/tensor.py` (#9) — `SymmetricTensor.random`'s `np.random.default_rng(seed)` seeding convention and its `ponytail:` note about real draws, which `random_isometry` deliberately departs from; `src/tenet/space.py` (#5) — `GradedSpace.degeneracy` returning `0` for an absent sector, and `ProductSpace.matches`; `src/tenet/array/dispatch.py` (#23) — the closed registration list, untouched; `tests/ops/test_embed.py` (#83) and `tests/ops/test_map.py` (#30) — the fixtures and the identity/containment assertions these criteria extend.
- **TensorKit.jl's constructors — the API being matched, with the exact names:** https://github.com/QuantumKitHub/TensorKit.jl , `src/tensors/linalg.jl` — `isometry([T=Float64,] codomain, domain)` / `isometry(codomain ← domain)` / `isometry(domain → codomain)`, "The isometry `t` then satisfies `t' * t = id(domain)` and `(t * t')^2 = t * t'`", enforced by `domain(t) ≾ codomain(t)` else `SpaceMismatch("domain and codomain are not isometrically embeddable: ...")`; `isomorphism` (`≅`, "no canonical choice … the current choice is such that `isomorphism(cod, dom) == inv(isomorphism(dom, cod))`"); `unitary`. All three fill blocks with `one!(b)` — deterministic, identity-shaped per block, which is what makes the `embed(identity(...))` reading exact rather than analogous.
- **TensorKit's Haar constructor:** `randisometry` / `randisometry!`, `src/auxiliary/random.jl` — `randisometry([rng], [T=Float64], dims::Dims{2})` implemented as `Q, = qr_compact!(Random.randn!(rng, A); positive = true)`, with `DimensionMismatch("cannot create isometric matrix with dimensions $dims; isometry needs to be tall or square")`; the `TensorMap` form `randisometry([rng], [TorA], codomain ← domain)` generated beside `rand`/`randn`/`randexp` in `src/tensors/tensor.jl`. **Note:** `randhaar` appears in a docstring header in that file but is not defined or exported anywhere — do not cite or copy that name.
- **Mezzadri, "How to generate random matrices from the classical compact groups", Notices of the AMS 54, 592 (2007)**, arXiv:math-ph/0609050 — the `Q * sign(diag(R))` correction, why plain QR of a Gaussian is not Haar, and the Ginibre draw for the unitary group. https://arxiv.org/abs/math-ph/0609050
- **quimb's `rand_iso` / `rand_uni`** — https://quimb.readthedocs.io/ — the dense equivalents, and the reason the per-sector-versus-dense distinction is worth a sentence in the docstring.
- `numpy.random.Generator.standard_normal` and `numpy.linalg.qr`: the two calls `random_isometry` makes; rectangular `ar.do("eye", n, m)` verified available on NumPy and JAX (autoray 0.10.1), and recorded here only as the rejected alternative.

**Out of scope**

- `unitary`, `isomorphism`, `random_unitary` as distinct functions. Equal-space cases of the two shipped ones, pinned by criteria instead.
- The fused, sector-wise containment case (a target where no single leg contains its partner but the fusion does). Refused; the construction exists (`eye(rows_c, cols_c)` per coupled sector) but names an arbitrary basis correspondence, and TensorKit's own `isomorphism` concedes the choice is non-canonical.
- A `SymmetricTensor.isometry` classmethod or method, and any `array/dispatch.py` registration. Constructors, not operations; the list stays closed (invariant 11).
- Backend-native RNG: `jax.random.PRNGKey`, torch generators, device-side draws. `to_backend` is the route, as it is for `random` and `zeros`.
- A `left`/`right`, `side=` or "co-isometry" variant. The adjoint of an isometry is one call.
- Any convenience pairing with `left_null` (a `complement_of` helper), or a CTM/VUMPS environment initializer. Those are `network/`'s.
- Non-Haar random isometries: a `rand_strength`-style near-identity perturbation, a structured or sparse isometry, a seedless global-RNG variant.
- Differentiating with respect to the *seed* or any reparametrized random isometry (a Cayley/exponential parametrization of the Stiefel manifold). That is an optimization technique, and `expm`/`qr` already supply its ingredients.
- PyTorch criteria.

**Acceptance criteria**

- `tenet.isometry(codomain, domain)` and `tenet.random_isometry(codomain, domain, seed=...)` exist in `src/tenet/ops/map.py`, are exported from `tenet` and `tenet.ops`, and carry legs `(*codomain OUT, *domain IN)` with `space`, `dual` and `name` taken from the arguments and only `side` set — `identity`'s documented stance, asserted.
- **`isometry` is `embed(identity(...))`, asserted rather than described:** the returned tensor is block-for-block equal (`0.0`) to `tenet.embed(tenet.identity(domain), target_legs)` for every fixture, so the implementation cannot drift from the composition it claims to be.
- **`W† W == identity(domain)`** exactly (`==` on the structural comparison, `0.0` on blocks) for Trivial, U(1), SU(2), fZ2 and a product symmetry, for a single-leg domain and a **two-leg** domain — the multi-leg case measured exact in the prototype and required, because it is where band interleaving could break the construction.
- **`W W†` is a projector:** `(W W†) @ (W W†) == W W†` to `1e-13`, and it equals `identity(codomain)` exactly when the containment is an equality — which doubles as the `unitary`-is-not-a-separate-function criterion.
- `isometry(legs, legs)` equals `identity(legs)` exactly, with every block object *identical* (`is`), inheriting `embed`'s allocation-light identity case.
- **Refusals are `embed`'s, verbatim and asserted on content:** a domain sector absent from the codomain, a domain sector with a larger degeneracy, a `dual` mismatch, a `side` disagreement, a provider mismatch and a leg-count mismatch each raise `ValueError` naming the axis, the sector and both degeneracies. A test records that these messages come from `_check_containment` and are not duplicated.
- **The per-leg restriction is documented and pinned:** a target where the fused spaces contain but no single leg does is refused, with a comment naming TensorKit's weaker `domain ≾ codomain` condition and the `ponytail:` note. Nobody discovers this from a wrong answer.
- **`random_isometry` satisfies `W† W == identity(domain)` to `1e-13`** for all four providers plus a product symmetry, on tall and on square coupled sectors, in `float64` and in `complex128`.
- **Haar-ness is measured, not assumed:** across many seeds, the mean of `|W_c[0,0]|²` per coupled sector matches `1/rows_c` to within the sampling error stated in the test, and — the criterion the sign fix exists for — a run *without* `Q * sign(diag(R))` fails it, recorded in a comment with the measured bias so the line cannot be deleted as redundant. Orthogonality alone is measured `4.4e-16`.
- **Complex draws are genuinely complex:** `random_isometry(..., dtype=complex128)` has a non-zero imaginary part in every block, and its per-sector eigenvalue-phase statistics are consistent with `U(n)` rather than `O(n)` — the test states which, with a comment recording that a real draw cast to complex would pass every other criterion here.
- Reproducibility: the same `seed` gives byte-identical blocks; different seeds differ; `seed=None` is documented as non-reproducible.
- **Structural refusal for `random_isometry`:** a coupled sector with `rows_c < cols_c` raises `ValueError` naming the sector and both dimensions, computed from `map_layout` before any draw — the *fused* condition, and the message says so, contrasting it with `isometry`'s per-leg one.
- Backend: both return NumPy blocks; `t.to_backend("jax")` preserves every criterion above to `1e-13`; `ops/map.py` gains no `import jax` and `test_only_pytree_module_imports_jax` passes unchanged, still asserting equality against `{"pytree.py", "ad.py"}`.
- **Cross-check with `left_null`** (the issue filed alongside): `adjoint(N) @ N` from `tenet.linalg.left_null` satisfies the same identity `isometry` guarantees, asserted in one test so the two definitions of "isometry" cannot drift apart.
- `array/dispatch.py` is unchanged and `tests/array/test_dispatch.py`'s closed-list assertion passes untouched.
- `ruff check`, `ruff format --check` and the full `pytest` suite pass.

**Depends on**

**#83** (`embed` and `_check_containment` — `isometry`'s entire implementation and every one of its refusals), **#30** (`identity`, whose blocks `isometry` embeds, and whose signature shape both functions follow; `compose`), #31 (`adjoint` — every criterion), #29 (`map_layout`, `MapLayout.shape` and `from_matrices` — `random_isometry`'s construction and its structural refusal), #9 (`SymmetricTensor.random`'s seeding convention, and the `ponytail:` note this deliberately departs from for complex dtypes), #5 (`GradedSpace`, `degeneracy` returning `0` for an absent sector), #6 (`Leg`, the independent `side`/`dual`), #8 (`TensorStructure`), #20 (`norm`), #41 (`tenet.pytree`, for the `to_backend` criteria).

Relates to the `left_null` / `right_null` issue filed alongside (its output is an isometry, cross-checked here) but does not depend on it; either can land first.
